### ISSUE TITLE
Make `hack/jenkins/e2e.sh` versioned with releases

### ISSUE THREAD

**Issue by ihmccreery** [kubernetes/kubernetes#17734]
Per problems related to #17180.  We need to be able to do large-scale refactors of tests without having versioning issues.  Because we drive tests from HEAD, the driver (`hack/jenkins/e2e.sh`) should be completely version-agnostic.

Alternatively, we can consider not driving tests from HEAD.  The question is just where we draw the line between "versioned" and "unversioned".

**Comment by karlkfi** [comment]
From #17180

> Well, I guess it's a good thing this came up, even if it did so in a somewhat painful way. Running unversioned test scripts against versioned tests is a good way to not be able to make any changes... I was unaware that was happening and didn't really have any way of knowing without access to your jenkins.

> The obvious solution is to run the tests on the branch you're testing. We're doing this in the Mesosphere CI, with some exception code that can override the CI scripts with ones from a specific repo/branch in case we need to override the scripts from another branch.

> Is there some reason the google jenkins can't do this?

**Comment by ixdy** [comment]
I think originally we'd intended to keep e2e.sh versioned with each release. For example, the comment on **kubernetes-e2e-gce-release-1.0** (still!) reads
```
  # Runs non-flaky tests on GCE on the release-1.0 branch,
  # sequentially. As a reminder, if you need to change the skip list
  # or flaky test list on the release branch, you'll need to propose a
  # pull request directly to the release branch itself.
```

I think the problem is that e2e.sh does too many things. It contains a lot of release/test infrastructure (which we'd probably like to keep at HEAD) as well as test selection (which makes more sense to be versioned). It seems unlikely that we'll want to continuously cherry-pick changes to e2e.sh into release branches as we muck with Jenkins, but we should still have the flexibility to change tests in master without needing all sorts of special casing for release branches. (Upgrade testing is another can of worms I'm not sure I want to think about right now.)

**Comment by karlkfi** [comment]
I can cherry-pick #17180 to the 1.1 release, but it would mean resolving a lot of conflicts and creating conflicts for later cherry-picks. I don't think that's a sustainable process.

Is there a reason you can't just run the scripts from the branch being tested? Then it doesn't matter what version you're testing, because the tests are implicitly versioned with the code.

Another other option, of course, is to extract the e2e tests to a new repo all together, but that requires an even stricter versioning scheme. That's probably better long term, tho.

In the single-repo world, tho, it's really only the conformance tests that should be run from a different branch than the code. That way we can have versioned conformance tests run against versioned and unversioned code. If we change them, it has to be done in a reverse compatible way. That was the original idea for the conformance branch, but it was never updated and no one else was testing conformance, so it became stale and unusable. We could start that over with 1.1, but it requires some additional work to make some sort of conformance test dashboard to report, otherwise there's no feedback loop.

**Comment by ixdy** [comment]
A test failure on #17735 made me remember that we're actually using in-branch e2e.sh on the PR builder already. This makes sense for changes to master (since we want to test anything new in e2e.sh), but it's pretty inconsistent for the release branches.

I think moving to per-branch e2e.sh makes the most sense, but I'm still not sure what we do for our version skewed builds.

Probably the easiest way to fix things (ignoring skew tests) is to merge the e2e.sh from master right before #17180 into the release branches, switch our Jenkins configs, and then explicitly delete the release job configs from master. (That way it's clear that fixes need to go into the release branches.)

Thoughts?
@kubernetes/goog-testing

**Comment by ihmccreery** [comment]
> I think moving to per-branch e2e.sh makes the most sense, but I'm still not sure what we do for our version skewed builds.

Yes; with the straw man proposal that version-skewed tests would be versioned on the latest version they are concerned with, (e.g. an upgrade test from 1.0 to 1.1 would be versioned with 1.1).

An alternative to versioning `e2e.sh` completely would be to clearly delineate what should be unversioned (e.g. "test infrastructure") and what should be versioned (e.g. test selection), and implement a separation of concerns accordingly.  #17180 actually gets us much closer to that; we could move to eliminating one-off skip-lists entirely, relying completely on `[labels]` in tests to slice and dice things.

There are enough considerations here that I think it merits a design doc.  @ixdy or @karlkfi, interested in writing one?

> Probably the easiest way to fix things (ignoring skew tests) is to merge the e2e.sh from master right before #17180 into the release branches, switch our Jenkins configs, and then explicitly delete the release job configs from master. (That way it's clear that fixes need to go into the release branches.)

SGTM; I can clean up skew tests once everything else is sane.  Again, a design doc would probably help us foresee problems before they come up; I'd be happy to review it with version-skew testing particularly in mind.

> Another other option, of course, is to extract the e2e tests to a new repo all together, but that requires an even stricter versioning scheme. That's probably better long term, tho.

This has come up in other spaces, too, particularly with conformance tests.  We often expect tests to be versioned with the thing they test, but while that definitely holds true for unit & integration tests, e2e tests may be a different beast.  @karlkfi want to file a separate issue about this?

**Comment by ixdy** [comment]
OK, so the basic idea is that we are moving to keep e2e.sh versioned on each branch as appropriate. I've started doing this in #17749, #17750, and #17751.

The basic idea is that we can enumerate all of the intended job configurations we want in the master-branch e2e.sh. When we cut a release, we'll similarly branch e2e.sh, which we know will be working for the release at that point. However, instead of adding additional job configurations to e2e.sh for the release branches (as we've done in the past), we'll instead keep reusing the branched configs.

For example, let's say on master we have:
* gce-parallel
* gce-slow
* gce-disruptive

We'll have Jenkins jobs named something like "kubernetes-e2e-gce-parallel-ci", "kubernetes-e2e-gce-slow-ci", etc, which will read the appropriate configurations from e2e.sh on the master branch.

For releases, we'll similarly have "kubernetes-e2e-gce-parallel-release", "kubernetes-e2e-gce-slow-release", etc jobs. These will read the same named configurations, except that they'll be coming from the release branch instead of master.

This helps solve the problem of e2e.sh exploding in size, as well as supporting a more maintainable process as we cut releases.

Note that with the PRs mentioned I'm only getting things back to the way they roughly were on Monday. We'll target infrastructure fixes for master/1.2 moving forward, keeping 1.0 and 1.1 infrastructure in a sort of maintenance-mode.

**Comment by ixdy** [comment]
There are still a few open questions:
* most of the GKE jobs should probably pull e2e.sh from the whatever release branch is current stable, since that's the version of Kubernetes it will be using, but there might be situations where we want to pull e2e.sh from a different location. 
* upgrade tests??? we ~~might~~ have tests that changed names between releases, so we potentially need to pull skip lists from multiple branches.

**Comment by ihmccreery** [comment]
Another open question:

- when we cut a release branch off of master, how do we do things like:
  - change job names,
  - remove old suites from the branch(es) we don't care about,
  - change jobs to point to the release branch?

cc @spxtr

**Comment by ihmccreery** [comment]
Another subtlety: there is still a bit of weirdness, since jobs sometimes run against the released version (e.g. v1.2.3), whereas e2e.sh usually pulls from HEAD of a release branch.  We could be even more draconian, and say that if you're running v1.2.3 tests, you pull e2e.sh *at* v1.2.3.

**Comment by wonderfly** [comment]
@ixdy @ihmccreery 

+1 for having versioned ```e2e.sh```, although it means we'd have to deal with (potentially) many merge conflicts when we backport infrastructure fixes. It helps to keep ```e2e.sh``` concise and clean.

I don't know why the kubernetes repo keeps so many stale branches, but ideally there'd be only two (or three) active branches where ```e2e.sh``` will be used - master (for CI jobs), and the latest (two) release branch(es) for release jobs. I assume when 1.2 is cut, 1.0 should be dropped. With that said, cherry-picking infra fixes wouldn't be as painful as we think.

> There are still a few open questions:
* most of the GKE jobs should probably pull ```e2e.sh``` from the whatever release branch is current stable, since that's the version of Kubernetes it will be using, but there might be situations where we want to pull e2e.sh from a different location.

Do you have any example of those situations?

> * upgrade tests??? we might have tests that changed names between releases, so we potentially need to pull skip lists from multiple branches.

Remind me what upgrade tests do again? I thought they launch a cluster with k8s at version A, upgrade the cluster to version B, and then run tests at version B. Isn't ```e2e.sh``` (and its skip lists) at version B exactly what we want? If some tests changed names from A to B, **their new names would have been reflected in the skip lists in ```e2e.sh``` at version B**.

> Another open question:
* when we cut a release branch off of master, how do we do things like:
  * change job names,

Instead of keeping switch-casing on full Jenkins job names, we could regulate job names to the format of, e.g., ```<meaningful-name>-<version_number>```, and switch-case on the ```<meaningful-name>``` part, which is unlikely to change from release to release. That way when we cut a release, we don't have to update job names in ```e2e.sh```. I don't think we'll ever go directly into a release branch and make changes there.
  * remove old suites from the branch(es) we don't care about,

In a normal git workflow, we would only create release branches off master, and changes to release branches should all come from master also (cherry-pick). If some tests become trashy, we remove them from ```e2e.sh``` at master and cherry-pick the change to active branches.
  * change jobs to point to the release branch?

I think this is the easiest - just update a Jenkins configuration, unless I missed something.

> Another subtlety: there is still a bit of weirdness, since jobs sometimes run against the released version (e.g. v1.2.3), whereas e2e.sh usually pulls from HEAD of a release branch. We could be even more draconian, and say that if you're running v1.2.3 tests, you pull e2e.sh at v1.2.3.

There won't be an active branch for a released version - it'll be only a tag. If we use ```e2e.sh``` from a tag, how do we patch it with infra fixes?

**Comment by ihmccreery** [comment]
> Instead of keeping switch-casing on full Jenkins job names, we could regulate job names to the format of, e.g., <meaningful-name>-<version_number>, and switch-case on the <meaningful-name> part, which is unlikely to change from release to release. That way when we cut a release, we don't have to update job names in e2e.sh.

I like this a lot.  It's actually similar to what I think we should do for upgrade tests, which are more complex than @wonderfly articulated above, e.g.:

    # kubernetes-upgrade-gke
    #
    # This suite:
    #
    # 1. launches a cluster at release/latest,
    # 2. upgrades the master to ci/latest,
    # 3. runs release/latest e2es,
    # 4. upgrades the rest of the cluster,
    # 5. runs release/latest e2es again, then
    # 6. runs ci/latest e2es and tears down the cluster.

I don't love this solution, but basically we can configure each branch to know how to configure itself for each step, and let the jobs pull from whatever branch they should pull from.  (See #17746)

> I don't think we'll ever go directly into a release branch and make changes there.

I don't think that's a reasonable assumption.  What if we realize there's some performance issue that we need a new suite for?  What if we decide we want to test a provider with a different configuration?  Ideally, all this would get worked out before the release gets branched, but that's not a guarantee.

> I think this is the easiest - just update a Jenkins configuration, unless I missed something.

What you missed is that "just update a Jenkins configuration" for ~60 jobs isn't something we want to do every three months.  It's not the worst, but it's *very* laborious, *very* prone to human error, and not transparent.  @spxtr for thoughts.

> There won't be an active branch for a released version - it'll be only a tag. If we use e2e.sh from a tag, how do we patch it with infra fixes?

Precisely—we can't.  So either the test infra on the release works, or it doesn't.  :smile:

**Comment by wonderfly** [comment]
> What you missed is that "just update a Jenkins configuration" for ~60 jobs isn't something we want to do every three months. It's not the worst, but it's very laborious, very prone to human error, and not transparent. @spxtr for thoughts.

We have to update ~60 jobs this time, after Jeff's refactoring, and I admit that updating them manually is painful and error prone. But why do we need to update them every three months? If we regulate the job names the way I suggested above, we don't have to update any existing job names. We do need to add a few new jobs for the new branch, but we'd have to do it anyways.

**Comment by ihmccreery** [comment]
Fair.  Hopefully @spxtr's work (reconciling jobs in branches with jobs in Jenkins) can address this better, though, so it's not laborious, not error-prone, and transparent.  Yay!

**Comment by wonderfly** [comment]
> I don't think that's a reasonable assumption. What if we realize there's some performance issue that we need a new suite for? What if we decide we want to test a provider with a different configuration?

Assuming we are developing at master, cut releases at master, and only cherry-pick critical features and bugfixes to release branches (that's what they are created for), if we find some performance issue or want to tweak some configuration, wouldn't master need these fixes and tweaks as well? But anyways, this debate seems to be beyond the issue we were dealing with, so I'll stop here. :smile: 

> Precisely—we can't. So either the test infra on the release works, or it doesn't.

This is something we'll need further discussion (see the internal bug tracker), but not a blocker for this particular issue imo.

> Hopefully @spxtr's work (reconciling jobs in branches with jobs in Jenkins) can address this better

Do you have a PR that I can refer to?

**Comment by ixdy** [comment]
> But why do we need to update them every three months? If we regulate the job names the way I suggested above, we don't have to update any existing job names.

We'll need to update the relevant release jobs to point to the correct branches. Though we could possibly automate this by having the jobs figure out the correct branch by looking at a file in GCS or git or something.

Your suggestion of switch-casing on features is basically exactly what I had in mind, you just articulated it better. In many ways the PR Jenkins is already doing what we want, since it already uses e2e.sh and all tests from a particular branch, rather than pulling configs from master. It just happens to use the same job name for all branches, rather than naming them differently.

**Comment by wonderfly** [comment]
> We'll need to update the relevant release jobs to point to the correct branches. Though we could possibly automate this by having the jobs figure out the correct branch by looking at a file in GCS or git or something.

Yes, for this time since it's the first time we started using versioned ```e2e.sh```. But new release jobs will be created with the correct names pointing to the correct branches that will never change util the branches are dead. What I was trying to say was we would not repeat this branch redirection every three months. We do it once and for all.

**Comment by ixdy** [comment]
@wonderfly we can either keep consistent job names (kubernetes-e2e-ci, kubernetes-e2e-release-latest, kubernetes-e2e-release-notquitelatest, etc), in which case we'll need to update their configs every release, or we can ping the release number in the job name as we've done up till now (kubernetes-e2e-ci, kubernetes-e2e-release-1.1, kubernetes-release-1.2, etc). The latter makes it very clear what's being tested, though it'll require spinning up/turning down jobs with each release.

**Comment by ihmccreery** [comment]
> it implies we need to update every single job except the ones for CI for every release

Unless I'm mistaken, this is easily resolved with an abstraction like a ENV variable, no?

Relatedly, I'm hating my life right now working on upgrades.  To put two upgrade suites in place (1.0 -> release/1.1 and 1.0 -> master), I have to coordinate PRs across three different branches: #18028, #18029, #18030.  I definitely don't think this is an acceptable solution in the long-run, (i.e. after Q1 2016).

I think the correct solution to this problem actually wasn't versioning `e2e.sh` and leaving it at that.  We instead need an extra layer of abstraction that decouples *how* from *what*.  `e2e.sh` right now describes *how* and *what*, (*what* is the massive `case` statement with configs, *how* is everything else).

@spxtr is working on a way to get the actual Jenkins configs under version control, so we should consider also moving all of the job-config stuff into those, (the *what*).  They should be unversioned (either at master, or, more likely, in a separate repository) and compatible with whatever version of `e2e.sh` (the *how*) they're pulling.

What do you all think?

**Comment by ihmccreery** [comment]
To be clear, *how* should never include the specifics of a job, just the properties that a certain "kind" of job should have (soak, upgrade, slow, etc.).

**Comment by wonderfly** [comment]
> Unless I'm mistaken, this is easily resolved with an abstraction like a ENV variable, no?

How do you apply a single ENV variable to multiple Jenkins jobs? Don't you need to update all of their configurations?

**Comment by karlkfi** [comment]
On a tangent, since you mentioned Jenkins configs (cc @spxtr): 
We run our Mesos CI tests from inside a kubernetes-mesos-test container. This has several benefits:

1. Dependency installation and versions are managed in version control in the Dockerfile
2. Nothing needs to be installed on the CI node (except docker).
3. All CI builds use the same test environment (except docker)
4. CI builds/tests can be run locally with a single command (provided your docker version matches the CI server)
5. Multiple versions of docker can be tested without modifying the test image (the socket and client binary are mounted)

These scripts are in [contrib/mesos/ci](https://github.com/kubernetes/kubernetes/tree/master/contrib/mesos/ci).
The kubernetes-mesos-test container is generated from [cluster/mesos/docker/test](https://github.com/kubernetes/kubernetes/tree/master/cluster/mesos/docker/test).

One caveat is that we only test with a single version of Golang at the moment. Changing that version requires having another Dockerfile and producing a docker image per Golang version.

**Comment by karlkfi** [comment]
Back on topic, I'm pretty convinced that the e2e tests should be in their own repo and versioned in branches/tags there, not mixed with the code they're testing. I've had a good experience doing this on CloudFoundry: [cf-acceptance-tests](https://github.com/cloudfoundry/cf-acceptance-tests), [cf-smoke-tests](https://github.com/cloudfoundry/cf-smoke-tests)

**Comment by ihmccreery** [comment]
> I'm pretty convinced that the e2e tests should be in their own repo and versioned in branches/tags there, not mixed with the code they're testing.

I've been considering the same thing.  I'm definitely convinced that it'd be a *lot* of work; I'm not convinced it would make the world much better.  @karlkfi can you say more about why they should be in their own repo?  What are the specific pros/cons?

**Comment by karlkfi** [comment]
Pros:

1) For one thing, the e2e suite isn't just a single suite any more. It's a lot of different suits with some common testing framework/library/infrastructure. So even just in that way, we're ripe for having multiple repos: test-tools, smoke, conformance, load, scale, destructive, etc. This would make versioning them harder, but maybe we really only need to version the conformance tests.
2) If you split out smoke and conformance from "the rest" then there's no such thing as "all of e2e". You just have a bunch of componentized tests in different repos that can all be run in parallel on different clusters. This would dramatically shorten the feedback cycle.
3) Another good reason is that versioning the tests makes it really easy to detect expectation changes, because reverse-incompatible changes to the tests will cause runs against existing code to fail. 
4) Having multiple test repos paves the way for having multiple core repos (like if the scheduler or certain controllers get extracted).
5) With the proper infrastructure, optional addons could have their own componentized e2e tests.
6) People can easily download and run the fast, non-destructive smoke tests against their own clusters every time they do an upgrade or deployment to verify the deployment health.

Cons:

1) One downside, is that you have to merge new features + new tests in two different PRs (3 if there's an integration repo) . So you need a way to specify which k8s PR satisfies which e2e PR (esp if they're not 1-to-1). One way to do this is with an integration repo that uses git sub-modules to link code & test repo revisions. You'd have to make a new integration repo branch/PR for every change in order to test code + test changes.
2) Updating the k8s code as Godeps into the test suite means you may need to bump it in most of the PRs. On the plus side, this means the new dependency change is tracked and not accidental. On the downside, it can make for Godep conflicts when master updates the same files as a PR.
3) Optional features like kube-dns make it necessary to track feature dependencies, which might require api groups in order to know what tests should pass on a specific cluster deployment.

**Comment by ihmccreery** [comment]
> 1) For one thing, the e2e suite isn't just a single suite any more. It's a lot of different suits with some common testing framework/library/infrastructure. So even just in that way, we're ripe for having multiple repos: test-tools, smoke, conformance, load, scale, destructive, etc. This would make versioning them harder, but maybe we really only need to version the conformance tests.

What do we gain from having multiple repos?  Having one repo with 12 packages sounds like a lot less of a headache than 12 repos each with their own package.

> 2) If you split out smoke and conformance from "the rest" then there's no such thing as "all of e2e". You just have a bunch of componentized tests in different repos that can all be run in parallel on different clusters. This would dramatically shorten the feedback cycle.

Again, I'm not sure this is a repo thing; I think it's a package/structure thing.

> 3) Another good reason is that versioning the tests makes it really easy to detect expectation changes, because reverse-incompatible changes to the tests will cause runs against existing code to fail.

Seems reasonable.  Right now we *do* version tests, right alongside code.  By pulling them out, we're effectively encouraging ourselves to decouple test version from code version.  I think this is the core of the argument for pulling them out, but I'm still not sure I have a clear sense of why it's better to decouple.  We *do* have the ability to run version-skewed code/test setups.  It's not very easy, but I'm not sure that separating the repos would make it easier. 

> 4) Having multiple test repos paves the way for having multiple core repos (like if the scheduler or certain controllers get extracted).

100% agree with this, and I think if we want to start extracting things, this may make sense.

> 5) With the proper infrastructure, optional addons could have their own componentized e2e tests.

Same as (1) and (2), not sure this relates to repo-management as much as test/suite management.

> 6) People can easily download and run the fast, non-destructive smoke tests against their own clusters every time they do an upgrade or deployment to verify the deployment health.

Same as (1), (2), and (5).

I agree with the cons you list.

Overall, I'm concerned that we're confusing repo-management with better e2e management.  I feel like pulling out into a separate repo isn't actually going to solve a lot of the pros you list, but please feel free to push back.  I'd rather focus on properly packaging different e2e suites, because I think that will have more positive impact without the negatives you list.

**Comment by karlkfi** [comment]
I do agree, that many benefits can be attained by better componentization via package management. That's actually a pre-requisite of pulling out to multiple repos anyway.

However, my experience with multiple components in the same repo suggests that the large a repo becomes, the harder it is to contribute to, regardless of how well packaged it is. The repo boundary is an administrative boon when it comes to things like determining ownership, getting reviews, versioning, releasing, productizing, documenting, having topical discussions with interested parties, avoiding merge conflicts, reducing team size, enforcing loose coupling, etc.

Instead of selling a solution of versioning or packaging or repo extraction, tho, we should probably concentrate on identifying the problems with the current solution that need addressing. Then we can make a better strategic strike plan.

**Comment by spxtr** [comment]
I agree with the general idea to move the test-infra and e2e stuff to a separate repository. I'm not convinced it's a good idea to have a separate repo per test, and I'm also not convinced it's high enough priority to be worth the effort.

It's possible to make a start on this by moving Jenkins-specific things (job configs, Jenkins docker image, etc), then e2e.sh, e2e.go, and so on, cleaning up as we go. That whole mess will need to be reworked at some point anyway, and this seems like a reasonable way to make a start at it.

**Comment by ixdy** [comment]
Given that `hack/jenkins/e2e.sh` is versioned with each branch now, I propose closing this issue. We can discuss moving test infra into its own repo in a separate issue, perhaps?

**Comment by ihmccreery** [comment]
This never got closed, but with https://github.com/kubernetes/kubernetes/issues/18122 done and https://github.com/kubernetes/kubernetes/pull/21122 in, I think `e2e.sh` is no longer in danger of being tightly coupled with versions.  Other issues remain, but I think we should close this (umbrella) issue.
